cmd/syncer/cmd/root.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+
+	nscontroller "github.com/kcp-dev/kcp/pkg/reconciler/namespace"
+	"github.com/kcp-dev/kcp/pkg/syncer"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/util/sets"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/component-base/logs"
+	"k8s.io/klog/v2"
+)
+
+const numThreads = 2
+
+func NewSyncerCommand() *cobra.Command {
+
+	options := NewDefaultOptions()
+	syncerCommand := &cobra.Command{
+		Use:   "syncer",
+		Short: "Synchronizes resources in `kcp` assigned to the clusters",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := options.validate(args)
+			return err
+		},
+	}
+	options.BindFlags(syncerCommand.Flags())
+	return syncerCommand
+}
+
+func (options *Options) validate(args []string) error {
+	//non-flag arguements
+	syncedResourceTypes := args
+	if len(syncedResourceTypes) == 0 {
+		syncedResourceTypes = []string{"deployments.apps"}
+		klog.Infoln("No resource types provided; using defaults.")
+	}
+	klog.Infof("Syncing the following resource types: %s", syncedResourceTypes)
+
+	// Create a client to dynamically watch "from".
+	if options.fromClusterName == "" {
+		return errors.New("--from_cluster is required")
+	}
+
+	fromConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: options.fromKubeconfig}, nil).ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	var toConfig *rest.Config
+	if options.toKubeconfig != "" {
+		var toOverrides clientcmd.ConfigOverrides
+		if options.toContext != "" {
+			toOverrides.CurrentContext = options.toContext
+		}
+
+		toConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: options.toKubeconfig},
+			&toOverrides).ClientConfig()
+		if err != nil {
+			return err
+		}
+	} else {
+		toConfig, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return err
+	}
```

Comment:
I think I would probably move this into Complete() too. Anything you're creating as variables in here could be stored as unexported fields in Options, then used in Run()

Reply:
yeah sure, that makes more sense.